provisioning/service/src/Authentication/ServiceConnectionStringBuilder.cs

```diff
@@ -93,20 +79,11 @@ private void Parse(string serviceConnectionString)
 
         private void Validate()
         {
-            if (string.IsNullOrWhiteSpace(SharedAccessKeyName))
-            {
-                throw new ArgumentException("Should specify SharedAccessKeyName");
-            }
+            Debug.Assert(!string.IsNullOrWhiteSpace(SharedAccessKeyName), "Should specify SharedAccessKeyName");
```

Comment:
We can't change these to asserts if it doesn't give feedback to the customer.

My earlier point was, if you thrown an ArgumentException or ArgumentNullException in a private/internal method based on the parameter names it knows about, that won't make sense to an end user who called a public API that doesn't have an argument with that name.

So, you either need to move exception throwing to the public API (it catches exceptions thrown earlier and rethrows specifically on an argument it knows has failed (like connection string).

Or you throw `InvalidOperationException` from deep in the code, and simply document that in the public API.

Which is better? I think it depends. Generally, I'd like to be as specific as possible, so the customer has a very actionable error, so I'd favor the former when possible.